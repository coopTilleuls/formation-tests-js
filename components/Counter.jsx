import {useMemo, useState} from 'react';
import translation from '../translation.json';
import {translateHelper} from '../helpers/translateHelper';

export const Counter = ({defaultValue = 1}) => {
  const [counter, setCounter] = useState(defaultValue);

  const decrement = () => setCounter(counter - 1);
  const increment = () => setCounter(counter + 1);
  const counterTranslation = useMemo(() => {
    return translateHelper.translate(translation, 'fr', 'counter', counter);
  }, [translation, counter]);

  return (
    <>
      <button data-testid="decrement" onClick={decrement}>-1</button>
      <span data-testid="counter">{counterTranslation}</span>
      <button data-testid="increment" onClick={increment}>+1</button>
    </>
  );
}
