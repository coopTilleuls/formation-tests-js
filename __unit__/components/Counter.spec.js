import {fireEvent, render, screen} from '@testing-library/react';
import '@testing-library/jest-dom'
import {Counter} from '../../components/Counter';

test('Should use one as default value', async () => {
  render(<Counter />)

  expect(screen.getByTestId('counter')).toHaveTextContent('1 produit');
});

test('Should use defaultValue props as default value', async () => {
  render(<Counter defaultValue="0"/>)

  expect(screen.getByTestId('counter')).toHaveTextContent('Rien');
});

test('Should increment counter on click to increment button', async () => {
  render(<Counter />)

  const incrementButton = screen.getByTestId('increment');
  fireEvent.click(incrementButton);
  fireEvent.click(incrementButton);
  fireEvent.click(incrementButton);

  expect(screen.getByTestId('counter')).toHaveTextContent('4 produits');
});

test('Should decrement counter on click to decrement button', async () => {
  render(<Counter />)

  const decrementButton = screen.getByTestId('decrement');
  fireEvent.click(decrementButton);
  fireEvent.click(decrementButton);

  expect(screen.getByTestId('counter')).toHaveTextContent('Rien');
});

test('Should increment counter on click to increment button (snapshot version)', async () => {
  const {container} = render(<Counter />)

  const incrementButton = screen.getByTestId('increment');
  fireEvent.click(incrementButton);
  fireEvent.click(incrementButton);
  fireEvent.click(incrementButton);

  expect(container).toMatchSnapshot();
});
