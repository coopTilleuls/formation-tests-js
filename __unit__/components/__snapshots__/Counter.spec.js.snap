// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should increment counter on click to increment button (snapshot version) 1`] = `
<div>
  <button
    data-testid="decrement"
  >
    -1
  </button>
  <span
    data-testid="counter"
  >
    4 produits
  </span>
  <button
    data-testid="increment"
  >
    +1
  </button>
</div>
`;
